docs/design-documents/20230228-connector-lifecycle-methods.md

```diff
@@ -0,0 +1,238 @@
+configurations will be populated (old contains the last active config values,
+new contains the latest config values).
+
+In the SDK we add 1 method to the `Source` and `Destination` interfaces:
+
+```go
+type Source/Destination interface {
+    ConfigChanged(ctx context.Context, oldConfig map[string]string, newConfig map[string]string) error
+}
+```
+
+Backwards compatibility with existing connectors is ensured through
+[`UnimplementedSource`](https://github.com/ConduitIO/conduit-connector-sdk/blob/edae165f49975c04fb5d9e549ea1987060068fc2/unimplemented.go#L49)
+and [`UnimplementedDestination`](https://github.com/ConduitIO/conduit-connector-sdk/blob/edae165f49975c04fb5d9e549ea1987060068fc2/unimplemented.go#L20),
+which all sources and destinations need to embed.
+
+#### Pros
+
+- Simpler design.
+- The logic for all three lifecycle events will probably be related, so it's
+  combined in one method.
+
+#### Cons
+
+- We risk having to add some data to one lifecycle event in the future and are
+  then forced to change it for the other ones as well (or pass `nil`).
+- The connector developer needs to figure out if the config was created, updated
+  or deleted.
+
+### Option 3 and 4
+
+We can combine the approaches described in option 1 and option 2 by choosing one
+for the connector protocol and the other for the SDK.
+
+#### Pros
+
+- Choose whatever approach makes sense in each part of the integration (protocol
+  and SDK).
+
+#### Cons
+
+- Looser connection between the protocol and SDK (different abstraction levels).
+
+### Common to all options
+
+The logic in Conduit will need to be adjusted so that lifecycle methods are
+triggered. When should the lifecycle methods be called? If we call them
+before `Configure` we risk that `Configure` later returns an error (i.e. config
+is invalid) but we still executed the initialization. On the other hand,
+if `Configure` was already called the lifecycle methods don't have to parse the
+config itself and they can know that the config is valid. That's why we should
+call lifecycle methods _after_ `Configure` and before `Open` (at that point the
+initialization already needs to be done).
+
+The lifecycle methods should be called at the following points:
+
+- Connector creation - should be called the first time the pipeline is started
+  after the connector was created.
+- Connector update - should be called the first time the pipeline is started
+  after the connector was updated (i.e. it was started once before already).
+- Connector deletion - should be called when the connector gets deleted. If it
+  fails, the error _should not_ prevent the deletion of the connector.
+
+We need to define which configuration gets passed to lifecyle methods. For this
+let's define that a connector configuration is "active" only after a lifecycle
+method was called with that configuration and no error was returned.
+Consequently, there are some edge cases we need to look out for:
```

Comment:
thanks for this section, it was helpful to understand the mechanics of the lifecycle events (particularly, why the `old` config is needed for deletion/updates)